text/0014-network-time.md

```diff
@@ -0,0 +1,681 @@
+
+![tickrate deprecations](0014-network-time/tickrate-deprecations.png)
+
+- `Network Tick Interval Sec` will now be given via the new `Tickrate` field.
+- `Receive Tickrate` is being removed. There is no replacement for this. MLAPI will now poll the receive queues at the beginning of every tick.
+- `Event Tickrate` will now be given via the new `Tickrate` field.
+- `Max Receive Events Per Tick Rate` will be removed. There will be no functionality to restrict the amount of events in MLAPI anymore as we move from an event to a snapshot based system.
+
+The default value for `Tickrate` will be `60` instead of `64`. A tickrate of 60 is more consistent with industry standard tickrates. Note that this rate is different than the default tickrate for FixedUpdate in Unity which is 50.
+
+### NetworkTime
+
+`NetworkTime` will be removed from MLAPI. As explained later in the RFC there are some flaws in the concept of a single `NetworkTime` so we will be replacing it with other concepts instead. For a quick fix of existing code instead of:
+```csharp
+var time = NetworkManager.Singleton.NetworkTime;
+```
+use
+```csharp
+var time = NetworkManager.Singleton.PredictedTime.Time;
+```
+
+## New Features
+
+### NetworkTime
+
+This RFC will introduce new features which can be used to interact with time in a networked manner.
+
+### Server Time
+A client receives regular state updates from the server containing data about the game state. This state could for instance be a position. Because time passes as packets travel between the server and clients this state is in a different position in time then the state of a locally controlled player character. `ServerTime.Time` can be used to interact with logic in this time space.
+
+For instance the following code could be used to displaying a moving platform which is aligned with the positions of all player objects from other players or server controlled NPCs:
+
+```csharp
+public class MovingPlatform : MonoBehaviour
+{
+    // Note moving in Update is only recommended for visual-only, non-networked objects.
+    public void Update(){
+        var positionX = Mathf.PingPong(NetworkManager.Singleton.ServerTime.Time)
+        transform.position = new Vector3(positionX, 0, 0);
+    }
+}
+```
+<img src="0014-network-time/platform_1.gif" width="640" height="360">
+
+### Predicted Time
+When moving your own player character locally, the character is ahead of time compared to all other NetworkObjects. This is the case because any changes to state will first need to be sent to the server and then back to all clients. `PredictedTime.Time` can be used to access this time. Most of the time `PredictedTime` should be used instead of `ServerTime` when implementing gameplay code.
+
+Example of using Predicted Time:
+
+```csharp
+public class NetworkChat : NetworkBehaviour
+{
+    [ServerRpc(RequireOwnership = false)]
+    SendChatMessageServerRpc(string message, float time)
```

Comment:
I ask you to change this and other absolute `time` to at least `double`s. You should never use `float` for absolute time.

For instance, if the app is storing time in float in seconds since app start time. 
The error in precision will be .976 msec after ~3 hours, and 7.81 msec after ~24 hours.

Source:
1. https://randomascii.wordpress.com/2012/02/13/dont-store-that-in-a-float/ 
2. https://twitter.com/ID_AA_Carmack/status/418158611664097280

Reply:
Agreed we should use double. Internally we are using a tick + float offset representation so we never have large absolute float values but I think it could still make sense to fully move do double. Since we have the chance to start fresh here with the API what would you recommend?
1. Make everything exposed a double. Both absolute and delta values.
2. Absolute values as doubles, delta values as floats.
3. Similar to the Unity Time API provide default as float but additional `AsDouble` API.

Reply:
3. is actually 2. Unity stores time as doubles only internally. It is just an API that truncates it to float (Time.time), and AsDouble gives it as-is.

Variant 2 looks right to me, but I don't have a strong opinion.